### ISSUE TITLE
All requests are using Queue Name instead of Queue URL

### ISSUE THREAD

**Issue by prokopyl** [ex-aws/ex_aws_sqs#8]
Right now, the ExAws.SQS library builds its request URIs by appending the queue name to the root path:

https://github.com/ex-aws/ex_aws_sqs/blob/185ec5b0ffa09ba2ca0f69ec73bda3a23f5c55f3/lib/ex_aws/sqs.ex#L279-L285

Not only is this undocumented behavior (which I'm surprised even works, considering it removes the IAM user ID from the Queue URL, though maybe for legacy reasons), but this is something [the AWS documentation explicitly says not to do](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-general-identifiers.html#queue-name-url):

> In your system, always store the entire queue URL exactly as Amazon SQS returns it to you when you create the queue (for example, https://sqs.us-east-2.amazonaws.com/123456789012/MyQueue). Don't build the queue URL from its separate components each time you need to specify the queue URL in a request because Amazon SQS can change the components that make up the queue URL. 

This is an issue for future compatibility with AWS services (as the documentation explicitly states this can change without warning), but it is also already an issue when using alternative AWS implementations such as [localstack](https://github.com/localstack/localstack) (which uses [ElasticMQ](https://github.com/softwaremill/elasticmq) under the hood).

As mentioned in #6, localstack/ElasticMQ uses the `/queue/` prefix before the Queue Name to build the the Queue URL (as allowed by the AWS documentation/specification, which is also coherent since they have no notion of IAM users, unlike the official implementation), causing it to only spit out errors when queried by ExAws (other clients work fine).

Crashing when using localstack is quite a big issue, since it prevents local development and testing, not only when using ExAws, but also frameworks like [broadway](https://github.com/plataformatec/broadway) which relies on ExAws.SQS.

Now, I am aware that switching from names to URLs would be a major, breaking change. I can however think of several solutions:

- If the queue parameter passed to `request`s passed to is not an URL, then call `GetQueueUrl` to retrieve the actual URL of the queue.
- Add an override to SQS request methods that accept a keyword list as a parameter instead of a string, which could be called as `ExAws.SQS.receive_message(queue_url: "[...]")` for instance. (this could be combined with the workaround above to preserve full compatibility, but opt-in to avoid the extra HTTP request).
- Make the breaking change and release a new major version. I do not know how acceptable this could be, however.

Of course there may be other solutions I did not see, but in any case if a decision is taken I am willing to submit some PRs to resolve this. (since I'm hacking around to make it work anyway right now :slightly_smiling_face: )

**Comment by benwilson512** [comment]
PR welcome :). There's no reason we can't release a 3.0 version.

**Comment by prokopyl** [comment]
Alright, I'll start working on a proper PR for this then!

**Comment by dnlserrano** [comment]
Let me know if I can help. I'd also benefit from this change. Thanks for raising this @prokopyl and thanks for this suite of libraries @benwilson512 and contributors. 🙌

**Comment by naps62** [comment]
@prokopyl @benwilson512 any updates on this?
I'm available to help on this. This issue is currently blocking me from using https://github.com/localstack/localstack

**Comment by jnatherley** [comment]
@naps62 Have you tried just prepending it with `"queue/<myqueue>"`

**Comment by naps62** [comment]
@jnatherley I did, and I get this when trying to send a message to it:

```
{:error,
             {:http_error, 400,
              %{
                code: "InvalidParameterValue",
                detail: "",
                message: "InvalidParameterValue; see the SQS docs.",
                request_id: "00000000-0000-0000-0000-000000000000",
                type: "Sender"
              }}}
```

It seems to me this is because the request looks like this:

```
%ExAws.Operation.Query{
  action: :send_message,
  content_encoding: "identity",
  params: %{
    "Action" => "SendMessage",
    "MessageBody" => "{\"foo\":\"bar\"}",
    "MessageGroupId" => "group_id"
  },
  parser: &ExAws.SQS.Parsers.parse/2,
  path: "/queue/my-queue.fifo",
  service: :sqs
}
```


But the SQS specs state that a `QueueUrl` argument is required

**Comment by TattdCodeMonkey** [comment]
@manelli @prokopyl  This is included in the v3.0.0 release
